### DISCUSSION TITLE
ensure identical chunk-hashes and chunk-ids on different platforms

### DISCUSSION THREAD

**Discussion by momesana** [webpack/webpack#15504] [Q&A]
Using webpack on different platforms to bundle an application results in different hashes and in some instances even different chunk ids. The respective portion of our webpack configuration looks as follows:
```ts
output: {
    filename: '[name].[contenthash].js',
    chunkFilename: '[name].[contenthash].chunk.js',
    ...
},
```
The hashes on linux and mac are identical. On windows they differ. We found out that the only thing inside the generated chunks that differs is the order of the keys of the module objects. Any idea on how to ensure identical hashes for identical sources regardless of the platform the code is bundled on?

**Comment by vankop** [comment]
hm.. we have such test cases.. could you provide reproducible repo? this maybe related to some non webpack core plugins.

**Reply by momesana** [reply]
I will try to create a separate repo for that then as the actual code in the repo is proprietary. Additionally the issue only seem to pertain to a handful of the chunks, notably one that involves the monaco-editor (that's the one where the id has changed too).

**Reply by momesana** [reply]
> hm.. we have such test cases.. could you provide reproducible repo? this maybe related to some non webpack core plugins.

The chunk where the id **and** the hash are different on windows is indeed caused by the [monaco-webpack-plugin](https://www.npmjs.com/package/monaco-editor-webpack-plugin). I'll create a minimal reproducable example for that and then check the other hash mismatches.

**Reply by momesana** [reply]
@vankop, here is a minimal example that reproduces the problem:  https://github.com/momesana/unstable-contenthash-when-using-monaco-webpack-plugin. In contrast to our own codebase (which is significantly larger and contains a bunch of other plugins) the name of the chunk is identical here and only the content hash differs.

I also reported this to the developers of the monaco-webpack-plugin, as this is likely caused by said plugin and not by webpack itself: https://github.com/microsoft/monaco-editor/issues/3021
 
 Once built under windows and linux (both via `npm run build`) there are only two sets of chunks with non-matching hashes:

**windows:**
```
594.990f43d1632048ca16e1.chunk.js // contains the monaco-editor chunk
594.990f43d1632048ca16e1.chunk.js.LICENSE.txtt
main.3e9baec4388838c4a569.js
main.3e9baec4388838c4a569.js.LICENSE.txt
```

**linux or mac:**
```
594.ae364e61732aa00da899.chunk.js // contains the monaco-editor chunk
594.ae364e61732aa00da899.chunk.js.LICENSE.tx
main.69cf259c81fb831bb4e6.js
main.69cf259c81fb831bb4e6.js.LICENSE.txt
```

The license files are identical and yield the same md5sum and the only difference in the `main.*.js` files once formatted is the hash of the referenced monaco-chunk:

```diff
1c1
< /*! For license information please see main.3e9baec4388838c4a569.js.LICENSE.txt */
---
> /*! For license information please see main.69cf259c81fb831bb4e6.js.LICENSE.txt */
6097c6097
<         594: "990f43d1632048ca16e1",
---
>         594: "ae364e61732aa00da899",
```

The chunks containing the monaco code itself however do deviate but only in the order in which two attributes appear in the code. Apart from that they are identical:
 ```diff
 1c1
< /*! For license information please see 594.ae364e61732aa00da899.chunk.js.LICENSE.txt */
---
> /*! For license information please see 594.990f43d1632048ca16e1.chunk.js.LICENSE.txt */
178c178
<     }, 9565: (e, t, i) => {
---
>     }, 4444: (e, t, i) => {
182c182
<         r.push([e.id, "/*---------------------------------------------------------------------------------------------\n *  Copyright (c) Microsoft Corporation. All rights reserved.\n *  Licensed under the MIT License. See License.txt in the project root for license information.\n *--------------------------------------------------------------------------------------------*/\n\n/* Uncomment to see lines flashing when they're painted */\n/*.monaco-editor .view-lines > .view-line {\n\tbackground-color: none;\n\tanimation-name: flash-background;\n\tanimation-duration: 800ms;\n}\n@keyframes flash-background {\n\t0%   { background-color: lightgreen; }\n\t100% { background-color: none }\n}*/\n\n.mtkcontrol {\n\tcolor: rgb(255, 255, 255) !important;\n\tbackground: rgb(150, 0, 0) !important;\n}\n\n.monaco-editor.no-user-select .lines-content,\n.monaco-editor.no-user-select .view-line,\n.monaco-editor.no-user-select .view-lines {\n\tuser-select: none;\n\t-webkit-user-select: none;\n\t-ms-user-select: none;\n}\n\n.monaco-editor .view-lines {\n\twhite-space: nowrap;\n}\n\n.monaco-editor .view-line {\n\tposition: absolute;\n\twidth: 100%;\n}\n\n.monaco-editor .mtkz {\n\tdisplay: inline-block;\n}\n\n/* TODO@tokenization bootstrap fix */\n/*.monaco-editor .view-line > span > span {\n\tfloat: none;\n\tmin-height: inherit;\n\tmargin-left: inherit;\n}*/\n", ""]);
---
>         r.push([e.id, "/*---------------------------------------------------------------------------------------------\n *  Copyright (c) Microsoft Corporation. All rights reserved.\n *  Licensed under the MIT License. See License.txt in the project root for license information.\n *--------------------------------------------------------------------------------------------*/\n.monaco-editor .lines-decorations {\n\tposition: absolute;\n\ttop: 0;\n\tbackground: white;\n}\n\n/*\n\tKeeping name short for faster parsing.\n\tcldr = core lines decorations rendering (div)\n*/\n.monaco-editor .margin-view-overlays .cldr {\n\tposition: absolute;\n\theight: 100%;\n}", ""]);
184c184
<     }, 4444: (e, t, i) => {
---
>     }, 9565: (e, t, i) => {
188c188
<         r.push([e.id, "/*---------------------------------------------------------------------------------------------\n *  Copyright (c) Microsoft Corporation. All rights reserved.\n *  Licensed under the MIT License. See License.txt in the project root for license information.\n *--------------------------------------------------------------------------------------------*/\n.monaco-editor .lines-decorations {\n\tposition: absolute;\n\ttop: 0;\n\tbackground: white;\n}\n\n/*\n\tKeeping name short for faster parsing.\n\tcldr = core lines decorations rendering (div)\n*/\n.monaco-editor .margin-view-overlays .cldr {\n\tposition: absolute;\n\theight: 100%;\n}", ""]);
---
>         r.push([e.id, "/*---------------------------------------------------------------------------------------------\n *  Copyright (c) Microsoft Corporation. All rights reserved.\n *  Licensed under the MIT License. See License.txt in the project root for license information.\n *--------------------------------------------------------------------------------------------*/\n\n/* Uncomment to see lines flashing when they're painted */\n/*.monaco-editor .view-lines > .view-line {\n\tbackground-color: none;\n\tanimation-name: flash-background;\n\tanimation-duration: 800ms;\n}\n@keyframes flash-background {\n\t0%   { background-color: lightgreen; }\n\t100% { background-color: none }\n}*/\n\n.mtkcontrol {\n\tcolor: rgb(255, 255, 255) !important;\n\tbackground: rgb(150, 0, 0) !important;\n}\n\n.monaco-editor.no-user-select .lines-content,\n.monaco-editor.no-user-select .view-line,\n.monaco-editor.no-user-select .view-lines {\n\tuser-select: none;\n\t-webkit-user-select: none;\n\t-ms-user-select: none;\n}\n\n.monaco-editor .view-lines {\n\twhite-space: nowrap;\n}\n\n.monaco-editor .view-line {\n\tposition: absolute;\n\twidth: 100%;\n}\n\n.monaco-editor .mtkz {\n\tdisplay: inline-block;\n}\n\n/* TODO@tokenization bootstrap fix */\n/*.monaco-editor .view-line > span > span {\n\tfloat: none;\n\tmin-height: inherit;\n\tmargin-left: inherit;\n}*/\n", ""]);
```

**Reply by Danielku15** [reply]
Just a hint which might be the reason: Windows uses by default `\r\n` as line endings and with git on windows it might convert the `\n` within the Git Repository to `\r\n` on the working directory. This would explain why the hashes are different. Have you tried configuring your repo to keep consistent line endings for your files so that you have still `\n` in your files within a Windows working directory? 

https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

I faced exactly this issue here in the WebPack repo because files are with `\n` and prettier complains constantly about `\r\n` that I remove the `\r`. I added locally the git attributes for all js files to work around this.

**Reply by momesana** [reply]
Hi @Danielku15,
this is one of the first things we checked. It is true that the line-endings may differ on windows and this becomes apparent if we check the checksums of the source files on windows vs. linux. They are indeed all different. Yet the issue described above only pertains to a very few chunks. In fact in the given example you can modify the code to not include the monaco-editor and the chunks will be identical regardless of the different line-endings/hashes on the different platforms. I've just done that in a branch called `without-monaco` in the repo above. Instead of the monaco editor there is a simple component that is lazily loaded in order to force a chunk to be created. The resulting chunks have indentical hashes on both platforms.

chunks on windows if monaco is not used:
```
15.03.2022  01:52    <DIR>          .
15.03.2022  01:52    <DIR>          ..
15.03.2022  01:52               501 130.1d06e52a0e4e1f5137e9.chunk.js
15.03.2022  01:52               324 88.f545f69adb5e4698fe5b.chunk.js
15.03.2022  01:52               188 index.html
15.03.2022  01:52           142.568 main.8081bffd652cc80daebb.js
15.03.2022  01:52               971 main.8081bffd652cc80daebb.js.LICENSE.txt
               5 File(s)        144.552 bytes
               2 Dir(s)  36.108.701.696 bytes free
```
and the same on linux:

```
130.1d06e52a0e4e1f5137e9.chunk.js
88.f545f69adb5e4698fe5b.chunk.js
index.html
main.8081bffd652cc80daebb.js
main.8081bffd652cc80daebb.js.LICENSE.txt
```
Plus, we know that the order of the some attributes differs in the generated files. I know that contenthash refers to content before it is being transpiled but I wouldn't be surprised if the imported files that make up the content of chunk are resolved or become available in a different order and cause this problem.
